libnetdata/popen/popen.c

```diff
@@ -89,118 +137,175 @@ static inline void convert_argv_to_string(char *dst, size_t size, const char *sp
+        if (posix_spawn_file_actions_addopen(&fa, STDOUT_FILENO, "/dev/null", O_WRONLY, 0)) {
+            error("POPEN: posix_spawn_file_actions_addopen() on stdout failed.");
+            // this is not a fatal error
+        }
+    }
+
+    if(!(error = posix_spawnattr_init(&attr))) {
         // reset all signals in the child
         sigset_t mask;
 
         if (posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETSIGMASK | POSIX_SPAWN_SETSIGDEF))
-            error("posix_spawnattr_setflags() failed.");
+            error("POPEN: posix_spawnattr_setflags() failed.");
+
         sigemptyset(&mask);
+
         if (posix_spawnattr_setsigmask(&attr, &mask))
-            error("posix_spawnattr_setsigmask() failed.");
-    } else {
-        error("posix_spawnattr_init() failed.");
+            error("POPEN: posix_spawnattr_setsigmask() failed.");
     }
+    else
+        error("POPEN: posix_spawnattr_init() failed.");
 
     // Take the lock while we fork to ensure we don't race with SIGCHLD
     // delivery on a process which exits quickly.
-    myp_add_lock();
+    netdata_popen_tracking_lock();
     if (!posix_spawn(&pid, command, &fa, &attr, (char * const*)spawn_argv, env)) {
         *pidptr = pid;
-        myp_add_locked(pid);
-        debug(D_CHILDS, "Spawned command: \"%s\" on pid %d from parent pid %d.", command_to_be_logged, pid, getpid());
-    } else {
-        myp_add_unlock();
-        error("Failed to spawn command: \"%s\" from parent pid %d.", command_to_be_logged, getpid());
-        if (flags & POPEN_FLAG_CREATE_PIPE) {
-            fclose(fp);
+        netdata_popen_tracking_add_pid_unsafe(pid);
+        netdata_popen_tracking_unlock();
+        debug(D_CHILDS, "POPEN: spawned command: \"%s\" on pid %d from parent pid %d.", command_to_be_logged, pid, getpid());
+    }
+    else {
+        netdata_popen_tracking_unlock();
+        error("POPEN: failed to spawn command: \"%s\" from parent pid %d.", command_to_be_logged, getpid());
+        if(flags & POPEN_FLAG_CREATE_PIPE) {
+            fclose(fp_child_stdin);
+            fp_child_stdin = NULL;
+            fclose(fp_child_stdout);
+            fp_child_stdout = NULL;
         }
         ret = -1;
     }
-    if (flags & POPEN_FLAG_CREATE_PIPE) {
-        close(pipefd[PIPE_WRITE]);
-        if (0 == ret) // on success set FILE * pointer
-            if(fpp) *fpp = fp;
+
+    if(flags & POPEN_FLAG_CREATE_PIPE) {
+        close(pipefd_stdin[PIPE_READ]);
+        close(pipefd_stdout[PIPE_WRITE]);
+
+        if (0 == ret) {
+            // on success set FILE * pointer
+            if (fpp_child_input) *fpp_child_input = fp_child_stdin;
+            if (fpp_child_output) *fpp_child_output = fp_child_stdout;
+        }
     }
 
-    if (!error) {
+    if(!error) {
         // posix_spawnattr_init() succeeded
         if (posix_spawnattr_destroy(&attr))
-            error("posix_spawnattr_destroy");
+            error("POPEN: posix_spawnattr_destroy() failed");
```

Comment:
I ran coverity on this PR and it is reporting:

```sh
*** CID 357335:  Resource leaks  (RESOURCE_LEAK)
/libnetdata/popen/popen.c: 281 in popene_internal()
275                 error("POPEN: posix_spawnattr_destroy() failed");
276         }
277     
278         if (posix_spawn_file_actions_destroy(&fa))
279             error("POPEN: posix_spawn_file_actions_destroy() failed");
280     
>>>     CID 357335:  Resource leaks  (RESOURCE_LEAK)
>>>     Variable "fp_child_stdout" going out of scope leaks the storage it points to.
281         return ret;
282     
283     error_after_posix_spawn_file_actions_init:
284         if (posix_spawn_file_actions_destroy(&fa))
285             error("POPEN: posix_spawn_file_actions_destroy() failed");
286     
```

Reply:
We have another one:

```sh
*** CID 357337:  Resource leaks  (RESOURCE_LEAK)
/libnetdata/popen/popen.c: 281 in popene_internal()
275                 error("POPEN: posix_spawnattr_destroy() failed");
276         }
277     
278         if (posix_spawn_file_actions_destroy(&fa))
279             error("POPEN: posix_spawn_file_actions_destroy() failed");
280     
>>>     CID 357337:  Resource leaks  (RESOURCE_LEAK)
>>>     Variable "fp_child_stdin" going out of scope leaks the storage it points to.
281         return ret;
282     
283     error_after_posix_spawn_file_actions_init:
284         if (posix_spawn_file_actions_destroy(&fa))
285             error("POPEN: posix_spawn_file_actions_destroy() failed");
286     
```

Reply:
I think this is by design. We need it to leak out of the function it is defined.
Do you see any problem with it?

Reply:
I agree with you, we can mark this as a false flag. 
I copied it here for we do no forget to mark it.

Reply:
If found it. It leaks both fp and the fpp params are NULL.

Reply:
I think I fixed it @thiagoftsm - can you pass it again over coverity?